File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.cpp
class Solution {
private:
    int binarySearch(vector<int> & nums,int key){
        int s = 0 ;
        int e = nums.size()-1;
        int mid = s + (e-s) / 2;
        while(s<=e){
            if(nums[mid]==key) return mid;
            else if (nums[mid] > key ) e = mid - 1 ;
            else s = mid + 1;
            mid = s + (e-s) / 2; 
        }
        return -1;
    }
public:
    int findPairs(vector<int>& nums, int k) {
        
        using pair = pair<int,int>;
        set<pair>s; // To maintain the uniqueness
        sort(nums.begin(),nums.end()); // Sorting (O(nlogn))
        
        for(int i=0;i<nums.size();i++){
            int i2 = binarySearch(nums,nums[i] + k);
            if(i2 != -1 && i!=i2) s.insert(make_pair(nums[i],nums[i2]));
        }
        
        return s.size();
    }
};

File: 0532-k-diff-pairs-in-an-array/0532-k-diff-pairs-in-an-array.py
class Solution:
    def binarySearch(self, nums, key):
        left, right = 0, len(nums) - 1
        
        while left <= right:
            mid = left + (right - left) // 2
            
            if nums[mid] == key:
                return mid
            elif nums[mid] > key:
                right = mid - 1
            else:
                left = mid + 1
        
        return -1
#-------------------------------------------------#    
    def findPairs(self, nums, k):
        unique_pairs = set()
        nums.sort()
        
        for i in range(len(nums)):
            i2 = self.binarySearch(nums, nums[i] + k)
            if i2 != -1 and i != i2:
                unique_pairs.add((nums[i], nums[i2]))
        
        return len(unique_pairs)
